### PR TITLE
Adding Enforce to platform

### PR DESCRIPTION
**Pull request by reyoung**
Basically from caffe2::logging.h, but only expose `PADDLE_ENFORCE`
interface.

### CODE REVIEW

File: paddle/platform/enforce.h [hunk 1]
```diff
@@ -0,0 +1,116 @@
+/*
```
Comment (jacquesqiao) [inline paddle/platform/enforce.h]: The copyright format is not right, please use  [this](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/gserver/layers/AddtoLayer.h#L1)

File: paddle/platform/enforce_test.cc [hunk 2]
```diff
@@ -0,0 +1,25 @@
+#include <gtest/gtest.h>
```
Comment (jacquesqiao) [inline paddle/platform/enforce_test.cc]: add copyright info.

File: paddle/platform/enforce.h [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+class EnforceNotMet : public std::exception {
+ public:
+  EnforceNotMet(const std::string& msg, const char* file, int fileline)
+      : file_(file), fileline_(fileline) {
+    std::ostringstream sout;
+    sout << msg << " at [" << file_ << ":" << fileline_ << "];";
+    all_msg_ = sout.str();
+  }
+
+  const char* what() const noexcept override { return all_msg_.c_str(); }
```
Comment (jacquesqiao) [inline paddle/platform/enforce.h]: why not just return a std::string

Comment (reyoung) [inline paddle/platform/enforce.h]: because it is the `std::exception` interface.

File: paddle/platform/enforce.h (hunk 1 above)
Comment (reyoung) [inline paddle/platform/enforce.h]: Done.

File: paddle/platform/enforce_test.cc (hunk 2 above)
Comment (reyoung) [inline paddle/platform/enforce_test.cc]: Done.

File: paddle/platform/enforce.h (hunk 3 above)
Comment (jacquesqiao) [inline paddle/platform/enforce.h]: got it! thanks~

File: paddle/platform/enforce.h [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+ private:
+  std::string all_msg_;
+  const char* file_;
+  int fileline_;
+};
+
+namespace details {
+
+inline void MakeStringInternal(std::ostringstream& stream) {}
```
Comment (wangkuiyi) [inline paddle/platform/enforce.h]: We should replace `MakeString` by `string::Printf` in the near future. I have a PR working on `string::Printf`.

Comment (reyoung) [inline paddle/platform/enforce.h]: Done, and wait for CI. After passing CI tests, this PR will be merged.

**Pull request merged**